Firestore/Swift/Source/ReferenceableObject/ReferenceableObjectManager.swift

```diff
@@ -0,0 +1,257 @@
+///
+///  To prevent writes of unmodified referenced objects, the manager compares checksums for the
+///  object being written.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public class ReferenceableObjectManager {
+  public static var instance = ReferenceableObjectManager()
+
+  static var settings = ReferenceableObjectManagerSettings()
+
+  private var db = Firestore.firestore()
+
+  private var objectCache = ReferenceableObjectCache()
+
+  public func save<T: ReferenceableObject>(object: T) async throws {
+    do {
+      let encoder = Firestore.Encoder()
+      let json = try encoder.encode(object)
+
+      if let docId = object.id {
+        guard let currentDigest = computeHash(obj: json),
+              await needsSave(object: object, currentDigest: currentDigest) else {
+          OSLog.storeByReference.debug("Object doesn't need to be saved")
+          return
+        }
+
+        try await db.collection(T.parentCollection()).document(docId).setData(json)
+        await objectCache.add(object: object, digest: currentDigest)
+      } else {
+        try await db.collection(T.parentCollection()).document().setData(json)
+      }
+
+      print("Save complete")
+    }
+  }
+
+  public func getObject<T: ReferenceableObject>(objId: String) async throws -> T? {
+    do {
+      // first check cache
+      if let cacheEntry = await objectCache.get(for: T.objectPath(objectId: objId)) {
+        return cacheEntry.object as! T
+      }
+
+      // get from db
+      let docRef = db.collection(T.parentCollection()).document(objId)
+      let doc = try await docRef.getDocument()
+      let obj = try doc.data(as: T.self)
+
+      // cache the doc since we just fetched it from store
+      if let jsonData = doc.data(),
+         let digest = computeHash(obj: jsonData) {
+        await objectCache.add(object: obj, digest: digest)
+      }
+
+      return obj
+    }
+  }
+
+  public func fetchObjects<T: ReferenceableObject>(type: T.Type) async throws -> [T] {
+    var foundObjects = [T]()
+    do {
+      let collectionRef = db.collection(type.parentCollection())
+      let docSnapshot = try await collectionRef.getDocuments()
+
+      for document in docSnapshot.documents {
+        let refObj = try document.data(as: T.self)
+        foundObjects.append(refObj)
+
+        let jsonData = document.data()
+        if let digest = computeHash(obj: jsonData) {
+          await objectCache.add(object: refObj, digest: digest)
+        }
+      }
+    }
+
+    OSLog.storeByReference.debug("fetchObjects found %ld objects", foundObjects.count)
+
+    return foundObjects
+  }
+
+  public func fetchObjects<T: ReferenceableObject>(predicates: [QueryPredicate]) async throws
+    -> [T] {
+    var query: Query = db.collection(T.parentCollection())
+
+    query = createQuery(query: query, predicates: predicates)
+
+    var foundObjects = [T]()
+    let snapshot = try await query.getDocuments()
+    for document in snapshot.documents {
```

Comment:
Maybe use a compactMap instead (like we do in most of our samples for mapping collections).

Reply:
This doesn't seem to work. I think the issue is that its within an async method.